### PR TITLE
gcc@4.9: Add RPATH

### PR DESCRIPTION
**Pull request by wangpeiwen**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Linuxbrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Linuxbrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
build args is only for mac, so I remove it on linux

### CODE REVIEW

**Comment by sjackman** [comment]
One of the 58 tests of `gmp` segfaults
```
/bin/bash: line 4: 24536 Segmentation fault      (core dumped) ${dir}$tst
FAIL: t-scan
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/191214594#L1421-L1422

**Comment by sjackman** [comment]
It's a known upstream bug. See https://gmplib.org/list-archives/gmp-bugs/2014-April/003451.html

Try adding the upstream patch to the `gmp` formula.
```ruby
patch do
  url "https://gmplib.org/repo/gmp/raw-rev/966737bd91ed"
  sha256 "4f4155d5e252ce457c1ad1b14aed1c8b56efe7402989efa254fae9040aa69917"
end
```

File: Formula/gmp@4.rb [hunk 1]
```diff
@@ -30,12 +30,14 @@ def install
...
-      ENV.m32
-      args << "--host=none-apple-darwin"
+    if OS.mac?
+      if MacOS.prefer_64_bit?
+        ENV.m64
+        args << "--build=x86_64-apple-darwin"
+      else
+        ENV.m32
+        args << "--host=none-apple-darwin"
+      end
```
Comment (sjackman) [inline Formula/gmp@4.rb]: To keep the patch smaller and make it easier for the weekly merge with Homebrew, please use:
```ruby
    if MacOS.prefer_64_bit?
      ENV.m64
      args << "--build=x86_64-apple-darwin" if OS.mac?
    else
      ENV.m32
      args << "--host=none-apple-darwin" if OS.mac?
    end
```

File: Formula/mpfr@2.rb [hunk 2]
```diff
@@ -34,12 +34,14 @@ def install
...
-      ENV.m32
-      args << "--build=none-apple-darwin"
+    if OS.mac?
+      if MacOS.prefer_64_bit?
+        ENV.m64
+        args << "--build=x86_64-apple-darwin"
+      else
+        ENV.m32
+        args << "--build=none-apple-darwin"
+      end
```
Comment (sjackman) [inline Formula/mpfr@2.rb]: And the same here.

**Comment by wangpeiwen** [comment]
done

**Comment by sjackman** [comment]
`gmp@4` works now. It's keg-only though, which means that it's not linked into `$HOMEBREW_PREFIX/lib`. As a result, `mpfr@2` isn't able to locate `libgmp.so.3` at run-time.
`error while loading shared libraries: libgmp.so.3: cannot open shared object file: No such file or directory`
https://travis-ci.org/Linuxbrew/homebrew-core/builds/191523876#L1475
A possible fix is to add `$(brew --prefix gmp@4)/lib` to the `RPATH` via `LD_RUN_PATH`. Something like (untested)…
```ruby
prepend_path "LD_RUN_PATH", Formula["gmp@4"].lib if OS.linux?
```
I'll have to think if there's a better fix.

**Comment by wangpeiwen** [comment]
I think we should build gcc and prerequisites in tree on linux which avoid those problems

File: Formula/gmp@4.rb [hunk 3]
```diff
@@ -8,6 +8,11 @@ class GmpAT4 < Formula
     sha256 "c259062b2b338e24503e0c0abd2d3467afd8b3ed48210aca7dc1206ea9ab4cc0" => :yosemite
   end
 
+  patch do
```
Comment (sjackman) [inline Formula/gmp@4.rb]: Please add a brief comment with the purpose of this patch.
```
# Fix make check error FAIL: t-scan
```

**Comment by wangpeiwen** [comment]
i make some tries and final build gcc 4.9 successfully
main changes:
+ add run path for gcc and libs by `-Wl,-rpath`
+ use ld because system's ld may be too old
+ add option without-sanitizer for some old kernel

**Comment by sjackman** [comment]
```
==> brew style gcc@4.9
== /home/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-core/Formula/gcc@4.9.rb ==
C: 72: 38: Missing space after #.

1 file inspected, 1 offense detected
==> FAILED
```
Please address this style issue.

File: Formula/cloog.rb [hunk 4]
```diff
@@ -38,6 +38,9 @@ def install
 
     args << "--with-osl=bundled" if build.head?
 
+    # add run path for linux
+    ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["isl@0.12"].opt_lib if OS.linux?
+    ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["gmp@4"].opt_lib if OS.linux?
```
Comment (sjackman) [inline Formula/cloog.rb]: I believe separating the two paths with a colon should work (untested).
```ruby
ENV.prepend "LDFLAGS", "-Wl,-rpath,#{Formula["isl@0.12"].opt_lib}:#{Formula["gmp@4"].opt_lib}" if OS.linux?
```

File: Formula/gcc@4.9.rb [hunk 5]
```diff
@@ -56,6 +56,7 @@ def osmajor
   option "with-profiled-build", "Make use of profile guided optimization when bootstrapping GCC"
   # enabling multilib on a host that can't run 64-bit results in build failures
   option "without-multilib", "Build without multilib support" if MacOS.prefer_64_bit?
+  option "without-sanitizer", "Build without libsanitizer" if OS.linux?
```
Comment (sjackman) [inline Formula/gcc@4.9.rb]: Did you observe this build failure, and with which version of the Linux kernel? I'd rather not include this build option unless necessary. Please include a link to a build log. You can use `brew gist-logs` to report build logs.
Please remove this line. We can address this in a different PR.

File: Formula/gcc@4.9.rb [hunk 6]
```diff
@@ -68,6 +69,7 @@ def osmajor
   depends_on "mpfr@2"
   depends_on "cloog"
   depends_on "isl@0.12"
+  depends_on "binutils" if OS.linux? #may need a modern ld
```
Comment (sjackman) [inline Formula/gcc@4.9.rb]: Did you observe a build failure, and with which version of `ld`?
Please include a link to a build log. You can use `brew gist-logs` to report build logs.
Please remove this line. We can address this in a different PR.

File: Formula/gcc@4.9.rb [hunk 7]
```diff
@@ -160,6 +164,22 @@ def install
...
+      ENV.prepend_path "LD_LIBRARY_PATH", Formula["isl@0.12"].opt_lib if OS.linux?
+      ENV.prepend_path "LD_LIBRARY_PATH", Formula["cloog"].opt_lib if OS.linux?
+      ENV.prepend_path "LD_LIBRARY_PATH", Formula["mpfr@2"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["gmp@4"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["libmpc@0.8"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["isl@0.12"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["cloog"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["mpfr@2"].opt_lib if OS.linux?
+      ENV.append "LDFLAGS", "-ldl" if OS.linux?
+      ENV["LD"] = Formula["binutils"].bin/"ld" if OS.linux?
```
Comment (sjackman) [inline Formula/gcc@4.9.rb]: Is this necessary? It should use the `ld` found in `PATH`, which should be the brewed `binutils`.

File: Formula/gcc@4.9.rb [hunk 8]
```diff
@@ -160,6 +164,22 @@ def install
...
+      args << "--with-boot-ldflags=" + ENV["LDFLAGS"] if OS.linux?
```
Comment (sjackman) [inline Formula/gcc@4.9.rb]: Please change  this line to
```ruby
    # Fix cc1: error while loading shared libraries: libisl.so.15
    args << "--with-boot-ldflags=-static-libstdc++ -static-libgcc #{ENV["LDFLAGS"]}" if OS.linux?
```
to make it match the corresponding line in `gcc.rb`

File: Formula/libmpc@0.8.rb [hunk 9]
```diff
@@ -26,6 +26,9 @@ def install
       "--with-mpfr=#{Formula["mpfr@2"].opt_prefix}",
     ]
 
+    # add run path for linux
+    ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["gmp@4"].opt_lib if OS.linux?
+    ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["mpfr@2"].opt_lib if OS.linux?
```
Comment (sjackman) [inline Formula/libmpc@0.8.rb]: Please combine these two lines as above.

File: Formula/gcc@4.9.rb [hunk 10]
```diff
@@ -160,6 +164,22 @@ def install
...
       end
 
+      # add run path for linux
+      # LD_LIBRARY_PATH and LDFLAGS is used for stage-1
+      # --with-boot-ldflags is for final gcc
+      ENV.prepend_path "LD_LIBRARY_PATH", Formula["gmp@4"].opt_lib if OS.linux?
+      ENV.prepend_path "LD_LIBRARY_PATH", Formula["libmpc@0.8"].opt_lib if OS.linux?
+      ENV.prepend_path "LD_LIBRARY_PATH", Formula["isl@0.12"].opt_lib if OS.linux?
+      ENV.prepend_path "LD_LIBRARY_PATH", Formula["cloog"].opt_lib if OS.linux?
+      ENV.prepend_path "LD_LIBRARY_PATH", Formula["mpfr@2"].opt_lib if OS.linux?
```
Comment (sjackman) [inline Formula/gcc@4.9.rb]: Please combine this into a single colon separated list.

File: Formula/gcc@4.9.rb [hunk 11]
```diff
@@ -160,6 +164,22 @@ def install
...
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["gmp@4"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["libmpc@0.8"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["isl@0.12"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["cloog"].opt_lib if OS.linux?
+      ENV.prepend "LDFLAGS", "-Wl,-rpath," + Formula["mpfr@2"].opt_lib if OS.linux?
```
Comment (sjackman) [inline Formula/gcc@4.9.rb]: Please combine this into a single colon separated list.

File: Formula/gcc@4.9.rb [hunk 12]
```diff
@@ -160,6 +164,22 @@ def install
...
+      ENV.append "LDFLAGS", "-ldl" if OS.linux?
```
Comment (sjackman) [inline Formula/gcc@4.9.rb]: It's odd that this line is necessary. Did you get a build failure without it?

File: Formula/gcc@4.9.rb (hunk 10 above)
Comment (sjackman) [inline Formula/gcc@4.9.rb]: If you set `RPATH` as below, do you also have to set `LD_LIBRARY_PATH`? I would have thought just setting `RPATH` would be enough.

Comment (sjackman) [inline Formula/gcc@4.9.rb]: If it is necessary to set both, stylistically please use
```ruby
if OS.linux?
  rpath = [Formula["foo"].opt_lib, …].join(":")
  ENV.prepend_path "LD_LIBRARY_PATH", rpath
  ENV.prepend "LDFLAGS", "-Wl,-rpath,#{rpath}"
end
```

File: Formula/gcc@4.9.rb [hunk 13]
```diff
@@ -148,6 +150,8 @@ def install
       args << "--enable-multilib"
     end
 
+    args << "--disable-libsanitizer" if build.without? "sanitizer"
```
Comment (sjackman) [inline Formula/gcc@4.9.rb]: Please remove this line. We can address this in a different PR.

**Comment by sjackman** [comment]
Please limit this PR just to what's necessary to get it working on the Linuxbrew CI on Ubuntu 12 and 14.
> add run path for gcc and libs by -Wl,-rpath

Keep this. Thanks!

> use ld because system's ld may be too old
> add option without-sanitizer for some old kernel

Remove these two items. We can address them in a different PR.

**Comment by wangpeiwen** [comment]
done

**Comment by sjackman** [comment]
Looking good, and getting closer.
`brew test mpfr@2` failed on Travis:
```
==> brew test mpfr@2 --verbose
==> /home/linuxbrew/.linuxbrew/bin/gcc-5 test.c -o test -lgmp -I/home/linuxbrew/.linuxbrew/Cellar/gmp@4/4.3.2/include -L/home/linuxbrew/.linuxbrew/Cellar/gmp@4/4.3.2/lib -lmpfr -I/home/linuxbrew/.linuxbrew/Cellar/mpfr@2/2.4.2/include -L/home/linuxbrew/.linuxbrew/Cellar/mpfr@2/2.4.2/lib
./test: error while loading shared libraries: libgmp.so.3: cannot open shared object file: No such file or directory
```
https://travis-ci.org/Linuxbrew/homebrew-core/builds/193285933#L1816

I find this error puzzling, as the path to `libgmp.so.3` should be stored in the `RPATH` of `libmpfr.so.2`. I'll have to look into it more.

**Comment by wangpeiwen** [comment]
```
==> brew test mpfr@2 --verbose
==> /home/linuxbrew/.linuxbrew/bin/gcc-5 test.c -o test -lgmp -I/home/linuxbrew/.linuxbrew/Cellar/gmp@4/4.3.2/include -L/home/linuxbrew/.linuxbrew/Cellar/gmp@4/4.3.2/lib -lmpfr -I/home/linuxbrew/.linuxbrew/Cellar/mpfr@2/2.4.2/include -L/home/linuxbrew/.linuxbrew/Cellar/mpfr@2/2.4.2/lib
./test: error while loading shared libraries: libgmp.so.3: cannot open shared object file: No such file or directory
```
because we use `-lgmp` which is unused and make test depend libgmp.so.3 directly without run path
remove it and everything is ok
PS: i have test it on mac

```
/usr/include/features.h:374:25: fatal error: sys/cdefs.h: No such file or directory
```
this may because ubuntu has no libc6-dev installed?
see [http://askubuntu.com/questions/470796/fatal-error-sys-cdefs-h-no-such-file-or-directory](http://askubuntu.com/questions/470796/fatal-error-sys-cdefs-h-no-such-file-or-directory)

**Comment by wangpeiwen** [comment]
i notice that rpath of dependency was added when building,
and i remove adding rpath in this PR
is there any other problem?
@sjackman

**Comment by sjackman** [comment]
Hi, @wangpeiwen. No particular problem, but I need to find some time to test that it works both with and without `glibc` installed, and with both `--env=super` and `--env=std`.

**Pull request closed without merging**